### ISSUE TITLE
Useful hooks/Utils for Contracts

### ISSUE THREAD

**Issue by carletex** [scaffold-eth/se-2#88]
When creating your app in SE, it's super handy to have hooks that loads the deployed contracts (hardhat_contracts / external_contracts).

In SE-2 we are using wagmi, but I think we could benefit of having some layer (hooks / utils) on top of wagmi to load these deployed contracts.

In SE-1, we have:
- useContractLoader (load a contract read / write)
- useContractReader (load & watch a contract variable)

This is what wagmi provides:
![image](https://user-images.githubusercontent.com/2486142/210976236-27c497c4-3e01-44cf-b135-2db3b12fa41a.png)

In my mind, what I want to do as a dev is something like this (rough sketch for the sake of the example):

```ts
// useContractLoader (or whatever name) will load the contract (read=no signer connected, write=signer connected).
// No need to pass any provider/signer params, since we have the wagmi hooks holding the state.
const yourContract = useContractLoader("YourContract");
await tx(yourContract.setPurpose("Hola", {value: utils.parseEther("0.001")}))

// useContractReader (or whatever name) will watch (all the time in localhost, rpc polling time in other networks) a variable/s within the contract.
const { purpose, userPurposeCounter } = useContractReader("YourContract", {config});
// Config could be an array of the things we want to read (with its params). Wagmi has useContractReads for multiple reads at the same time. E.g config:
const config = {
 ["purpose", []]
 ["userPurposeCounter", ["0xaddresss"]]
}
```

We'll want to leverage as much as possible on the TS auto-completions for the contracts.

**Comment by Naim-Bijapure** [comment]
@carletex  on the previous nextjs template have tried this . 
https://github.com/Naim-Bijapure/scaffold-eth-next/blob/main/packages/next-ts/hooks/useAppLoadContract.ts

created an useAppLoadContract hook which takes a contract name as an argument and returns its instance  and types for typescript autocomplete ( inspired from SE typescript). 

 there using useContract hook from wagmi to return contract instance. 
 also there need to configure contract abi files paths in a constant file first. 
 
what do you think? 
should I start digging into this?

**Comment by technophile-04** [comment]
Really love the idea of "useful hooks/Utils for Contracts" much needed !!! 

I have been thinking and digging about it and here are some of my thoughts :

## Hooks idea : 
 ### 1) wrapper for `useContractRead` or `useContractWrite` : 
 Reason: every time we use `useContractRead` or `useContractWrite` we always need to pass `addressOrName`, `abi`checkout docs [here](https://0.6.x.wagmi.sh/docs/hooks/useContractRead#configuration) instead what we can do create wrappers which fills in this value by just passing contract name : 
 ```js
 const { result, isLoading, isError } = useScaffoldRead<YourContract>({
  contractName: "YourContract",
  functionName: "getCount",
  watch: true,
});

//---- Rough Implementation :
const useScaffoldRead = ({ contractName, ...rest }) => {
  const contractDetails = useWagmiContractOptions(contractName);
  return useContractRead({
    ...conteractDetails,
    ...rest,
  });
};

 ```
 similar for `useContractWrite` also we should handle types for better auto-completion.
 Inspiration was taken from [here](https://github.com/holic/web3-scaffold/blob/f4f018df388d2838f895c7a381eafe19e299c757/packages/app/src/contracts.ts#L21)
 
 ### 2) useWagmiContractOptions hook : 
 We already have this util [getDeployedContract](https://github.com/scaffold-eth/se-2/blob/b3efaa60d21660983ea20a32afb6509ae8788fc0/packages/nextjs/components/scaffold-eth/Contract/utilsContract.tsx#L20) but maybe we can convert it to hook so that we don't need to pass it chain Id also it will be utilized by useContractRead and useContractWrite with those 2 value (`addressOrName`, `abi`)
 
 
##  Regarding `useContractLoader` : 
The idea is really great, but if the hook return contract Instance even Naim suggested the same approach then its just good to have a hook,I feel that making `write` call wtih that instance is more declarative which I don't like,  I like the wagmi's approach of `write` because it handles other state's too like `isLoading`, `error` etc.  but its good to have. 

PS : These are all my personal views and very raw drafts nd we can ignore some of them too

**Comment by carletex** [comment]
Thanks @Naim-Bijapure and @technophile-04 !! Great suggestions.

I think Shiv is drafting a PR with a first iteration on this. We can play with it and take it from there.

**Comment by carletex** [comment]
This was merged on #93 

We'll iterate... but let's close this for now.
